service.py

```diff
@@ -0,0 +1,263 @@
+    rows = []
+    logger.debug('Opening {} containing hathiTrust records'.format(localFile))
+
+    try:
+        hathiFile = open(localFile, newline='')
+    except FileNotFoundError:
+        logger.error('Could find file at path {}'.format(localFile))
+        raise ProcessingError('loadLocalCSV', 'Could not open local CSV file')
+
+    with hathiFile:
+        hathiReader = csv.reader(hathiFile)
+        for row in hathiReader:
+            if row[0] == 'htid':
+                continue
+            rows.append(row)
+    logger.debug('Loaded {} rows from {}'.format(str(len(rows)), localFile))
+    return rows
+
+
+def fetchHathiCSV():
+    return None
+
+
+def fileParser(fileRows, columns):
+    """Iterates through parsed CSV rows to return a list of records formatted
+    to comply with the SFR data format. As a requirement of this a dict of
+    country codes/country names is generated here for use as a lookup table.
+
+    Arguments:
+    fileRows -- list of parsed CSV rows
+    columns -- list of column names that correspond to the CSV file
+
+    Output:
+    outcomes -- list of processing result tuples from the rowParser() function.
+    These are either 'success' or 'failure' and provide resiliency in logging
+    errors while allowing this method to continue iteration over rows
+    containing errors
+    """
+
+    logger.info('Parsing rows retrieved from source file')
+
+    logger.debug('Loading country codes from XML file')
+    countryCodes = loadCountryCodes()
+
+    outcomes = []
+    for row in fileRows:
+        try:
+            outcomes.append(rowParser(row, columns, countryCodes))
+        except ProcessingError as err:
+            outcomes.append(('failure', err.source, err.message))
+
+    return outcomes
+
+
+def rowParser(row, columns, countryCodes):
+    """Parse single HathiTrust item entry (corresponding to an item-level
+    record in the SFR model) into the SFR data model and pass the resulting
+    object to Kinesis for introduction into the SFR data pipeline.
+
+    This method is a manager that handles methods around a HathiRecord object.
+    Each method creates/enhances a part of the SFR metadata object, allowing
+    for the object to both be built up and its components easily treated
+    as seperate components if necessary
+
+    Arguments:
+    row -- list of fields from the HathiTrust source CSV file
+    columns -- list of columns that corresponds to the source row
+    countryCodes -- dict of country code and name translations
+
+    Output: None, writes resulting work record to a Kinesis stream
+    """
+
+    logger.info('Reading entry for HathiTrust item {}'.format(row[0]))
+
+    logger.debug('Generating source dict from row and column names')
+    # This quickly builds a dictionary with column names that can be used to
+    # retrieve specific values
+    hathiDict = dict(zip(columns, row))
+    # Generate a hathi record object with the source dict
+    hathiRec = HathiRecord(hathiDict)
```

Comment:
Would you instantiate a `HathiRecord` without building it? My first reaction is that `buildWork()`, `buildInstance()` etc. could just be part of construction. But maybe you want the option of logging the steps (and you don't want to build that into the `HathiRecord` object)?   

I'm not sure what the trade-offs are. There's kind of a mix of encapsulation (like `buildWork()`) and transparency (like directly manipulating `hathiRec.work.instances`). I wonder if it could be clearer?

Reply:
I think it can be and personally I prefer to keep most of this logic within the class itself/outside of the general handler (division of responsibility and all that). I added a new function `buildDataModel` that better encapsulates the logic here and which feels more clear to me. It is a little less transparent but breaks it down it more logical level of details so someone reading the code can "descend" to each level as necessary to get more detail.

You could certainly create an empty `HathiRecord` object and then build the source dict, perhaps from multiple sources (I'm thinking if we perhaps end up using the API to get more detailed info from Hathi on top of what's available in the CSV files).